core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,380 @@
+	}
+}
+
+// MergeFilter will update the filter with the new topics and addresses.
+// Clients may chose to MergeFilter and then replay in order to ensure desired logs are present.
+func (lp *LogPoller) MergeFilter(topics []common.Hash, address common.Address) {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	lp.addresses[address] = struct{}{}
+	// [[A, B], [C]] + [[D], [], [E]] = [[A, B, D], [C], [E]]
+	for i := 0; i < len(topics); i++ {
+		if lp.topics[i] == nil {
+			lp.topics[i] = make(map[common.Hash]struct{})
+		}
+		lp.topics[i][topics[i]] = struct{}{}
+	}
+}
+
+func (lp *LogPoller) filterAddresses() []common.Address {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var addresses []common.Address
+	for addr := range lp.addresses {
+		addresses = append(addresses, addr)
+	}
+	sort.Slice(addresses, func(i, j int) bool {
+		return bytes.Compare(addresses[i][:], addresses[j][:]) < 0
+	})
+	return addresses
+}
+
+func (lp *LogPoller) filterTopics() [][]common.Hash {
+	lp.filterMu.Lock()
+	defer lp.filterMu.Unlock()
+	var topics [][]common.Hash
+	for idx := 0; idx < len(lp.topics); idx++ {
+		var topicPosition []common.Hash
+		for topic := range lp.topics[idx] {
+			topicPosition = append(topicPosition, topic)
+		}
+		sort.Slice(topicPosition, func(i, j int) bool {
+			return bytes.Compare(topicPosition[i][:], topicPosition[j][:]) < 0
+		})
+		topics = append(topics, topicPosition)
+	}
+	return topics
+}
+
+// Replay signals that the poller should resume from a new block.
+func (lp *LogPoller) Replay(fromBlock int64) {
+	lp.replay <- fromBlock
+}
+
+func (lp *LogPoller) Start(parentCtx context.Context) error {
+	return lp.StartOnce("LogPoller", func() error {
+		ctx, cancel := context.WithCancel(parentCtx)
+		lp.ctx = ctx
+		lp.cancel = cancel
+		go lp.run()
+		return nil
+	})
+}
+
+func (lp *LogPoller) Close() error {
+	return lp.StopOnce("LogPoller", func() error {
+		lp.cancel()
+		<-lp.done
+		return nil
+	})
+}
+
+func (lp *LogPoller) Ready() error {
+	return nil
+}
+
+func (lp *LogPoller) Healthy() error {
+	return nil
+}
+
+func (lp *LogPoller) run() {
+	defer close(lp.done)
+	tick := time.After(0)
+	var start int64
+	for {
+		select {
+		case <-lp.ctx.Done():
+			return
+		case fromBlock := <-lp.replay:
+			lp.lggr.Warnw("Replay requested", "from", fromBlock)
+			start = fromBlock
+		case <-tick:
+			tick = time.After(utils.WithJitter(lp.pollPeriod))
+			if start == 0 {
+				lastProcessed, err := lp.orm.SelectLatestBlock(pg.WithParentCtx(lp.ctx))
+				if err != nil {
+					if !errors.Is(err, sql.ErrNoRows) {
+						lp.lggr.Warnw("unable to get starting block", "err", err)
+						continue
+					}
+					// Otherwise this is the first poll _ever_ on a new chain.
+					// Only safe thing to do is to start at finality depth behind tip.
+					latest, err := lp.ec.BlockByNumber(context.Background(), nil)
+					if err != nil {
+						lp.lggr.Warnw("unable to get latest for first poll", "err", err)
+						continue
```

Comment:
Same as above, this sort of blocks the poller. But this could still be a softer error, since we can hope that we can eventually be able to reach a node to give us the correct blocknumber.

Reply:
yeah it is intentional that we block the poller in this case